### PR TITLE
Base Template Generator

### PR DESCRIPTION
**Pull request by mohamedashraf-eng**
**Added a utility to generate templates for easy development.**  

Features:  

- Can generate multiple .c .h files with base template.
- Ability to integrate multiple licenses.  

Future updates:  

- Flexible template adding.
- Full template generate with functions/services and variables documenting.
- Json serialization. 

Working:  

`pip install -r requirements.txt`
`add files to generate in gen_files.txt`
`python3 main.py`

### CODE REVIEW

**Comment by autosarzs** [comment]
Lets wait for a little bit until we defien the final structure of the build system, then we will use this tool not only to create the files but also to create the CMake files with proper structure for both build and test

File: tools/utils/base_generator/main.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        with open(f'{GENERATED_FILES_DIR}/{file}' + '.h', 'w') as f:
+            f.write(h_file_license_headerx)
+            f.write(h_file_header)
+            f.write(f"\n#if !defined( __{file.upper()}_H__ )\n#define ( __{file.upper()}_H__ ) \n")
+            f.write("\n#if defined(cplusplus) || defined(c_plus_plus)\nextern \"c\" {\n")
+            f.write("#endif /* defined(cplusplus) || defined(c_plus_plus) */ \n\n")
+            f.write(h_file_headers)
+            f.write("\n#if defined(cplusplus) || defined(c_plus_plus)\n}\n")
+            f.write("#endif /* defined(cplusplus) || defined(c_plus_plus) */")
+            f.write(f"\n\n#endif /* ( __{file.upper()}_H__ ) */")
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:236]: I suggest adding empty line at the end.

File: tools/utils/base_generator/main.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+@file: main.py
+@author: MoWx 
+@brief:
+@copyright: 
+
+@attention:
+'''
+import os
+
+import os
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:11]: Repeated line 
```suggestion
import os
```

File: tools/utils/base_generator/main.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            f.write(f"\n\n#endif /* ( __{file.upper()}_H__ ) */")
+            
+if __name__ == '__main__':
+    file_names = []
+        
+    with open('gen_files.txt', 'r') as file:
+        for g_files in file:
+            file_names.append(g_files.strip())
+        
+    write_to_files(file_names)
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:245]: Please, Add an empty line at the end and apply it to all files.

File: tools/utils/base_generator/main.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                for segment in segments:
+                    num_spaces = max_line_length - len(segment) + 3
+                    formatted_segment = '** {}{} **'.format(segment.strip(), ' ' * num_spaces)
+                    formatted_segments.append(formatted_segment)
+                formatted_line = '\n'.join(formatted_segments)
+                license_header.append(formatted_line)
+                
+    return '\n'.join(license_header) + '\n'
+            
+c_file_license = None
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:45]: How did you hard code this? It makes the get_license_header function meaningless.

File: tools/utils/base_generator/main.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+**                                                                            **
+**  DATE         :                                                            **
+**                                                                            **
+**  VARIANT      :                                                            **
+**                                                                            **
+**  PLATFORM     :                                                            **
+**                                                                            **
+**  AUTHOR       :                  	                                      **
+**                                                                            **
+**  VENDOR       :                                                            **
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:84]: I believe we should initialize those data fields that are empty.

File: tools/utils/base_generator/main.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+/*                                   Local Function Definition                           */
+/*****************************************************************************************/
+
+/*****************************************************************************************/
+/*                                   Global Function Definition                          */
+/*****************************************************************************************/
+'''
+
+c_file_service_header = ''' 
+/************************************************************************************
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:143]:
```suggestion
c_file_service_header = ''' /************************************************************************************
```

File: tools/utils/base_generator/main.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                license_header.append(formatted_line)
+                
+    return '\n'.join(license_header) + '\n'
+            
+c_file_license = None
+if c_file_license is not None:
+    c_file_license_header = get_license_header(c_file_license, 71)
+
+    c_file_license_headerx = '''
+/*******************************************************************************
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:50]: To not start with empty line.
```suggestion
    c_file_license_headerx = '''/*******************************************************************************
```

File: tools/utils/base_generator/main.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+**  LICENSE:                                                                  **
+{}
+/*******************************************************************************
+    '''.format(c_file_license_header)
+else:
+    c_file_license_headerx = '''
+/*******************************************************************************
```
Comment (Mina-Karam) [inline tools/utils/base_generator/main.py:57]: To not start with empty line.
```suggestion
    c_file_license_headerx = '''/*******************************************************************************
```

**Review by Mina-Karam — COMMENTED** [review]
First, Great job here. I have some points I suggest to cover

1- The `requirments.txt` file is now empty, so no need for it.
2- The `autosarz.txt` is not used; you already hard-coded it into the Python script.
3- Rename the `main.py` to `template_generator.py`.
4- I suggest not to have `gen_files.txt`; instead, pass the file names as parameters to the script.
5- I suggest adding a function called `usage ` if someone calls the script in the wrong way to show how to use it.
For Ex: 
```python
def usage():
    print('template_generator.py script usage')
    print('template_generator.py -l <file names> -o <Output location>')
    sys.exit(1)
```
6- Setting the template's initialize value to something like the file name, version, and so on.
7- We use Mozilla Public License Version 2.0, not MIT, and it has already been added; there is no need for each file.

Thanks again for your great efforts.

**Comment by mohamedashraf-eng** [comment]
@Mina-Karam 
First of all thanks for your great effort reviewing this and for the comments I will make sure to add all of this next version. I made this as base idea I agree it is not that efficient, after we identify the full template that we will use, I will add version 2.

**Pull request closed without merging**